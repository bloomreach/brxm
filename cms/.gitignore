*.iml
*.log
*.tmp
/.idea/
.classpath
.project
.settings
.DS_Store
target/
/api/node/
/api/node_modules/
/api/repository/
/api/version/
/api/workspaces/
/automatic-export/repository/repository/
/automatic-export/repository/version/
/automatic-export/repository/workspaces/
/console/frontend/node/
/console/frontend/node_modules/
/editor/repository/workspaces/
/engine/node/
/engine/node_modules/
/richtext/frontend/repository/
/richtext/frontend/version/
/richtext/frontend/workspaces/
/test/repository/
/test/version/
/test/workspaces/
/workflow/frontend/repository/
/workflow/frontend/version/
/workflow/frontend/workspaces/
