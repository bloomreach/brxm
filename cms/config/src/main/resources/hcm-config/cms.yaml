definitions:
  config:
    /hippo:configuration/hippo:frontend/cms:
      jcr:primaryType: frontend:application
      frontend:defaultcluster: cms-static
