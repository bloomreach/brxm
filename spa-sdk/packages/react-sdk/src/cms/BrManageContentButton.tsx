/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { ManageContentButton, TYPE_MANAGE_CONTENT_BUTTON } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';

/**
 * The button component that opens for editing a content.
 */
export class BrManageContentButton extends React.Component<ManageContentButton> {
  static contextType = BrPageContext;
  context: React.ContextType<typeof BrPageContext>;

  render() {
    return this.context?.isPreview()
      ? <BrMeta meta={this.context.getButton(TYPE_MANAGE_CONTENT_BUTTON, this.props)} />
      : null;
  }
}
