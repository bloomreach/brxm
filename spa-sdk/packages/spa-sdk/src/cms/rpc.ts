/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { EmitterMixin, Emitter } from '../emitter';

export const RpcClientService = Symbol.for('RpcClientService');
export const RpcServerService = Symbol.for('RpcServerService');

type Callable<T = any, U extends unknown[] = any[]> = (...args: U) => T;

const TYPE_EVENT = 'brxm:event';
const TYPE_RESPONSE = 'brxm:response';
const TYPE_REQUEST = 'brxm:request';
const STATE_FULFILLED = 'fulfilled';
const STATE_REJECTED = 'rejected';

export interface Event {
  type: typeof TYPE_EVENT;
  event: string;
  payload?: any;
}

export interface Request {
  type: typeof TYPE_REQUEST;
  id: string;
  command: string;
  payload: any[];
}

export interface Response {
  type: typeof TYPE_RESPONSE;
  id: string;
  state: typeof STATE_FULFILLED | typeof STATE_REJECTED;
  result: any;
}

export type Message = Event | Request | Response;
export type Procedures = Record<string, Callable>;
export type Events = Record<string, any>;

export interface RpcClient<TProcedures extends Procedures, TEvents extends Events> extends Emitter<TEvents> {
  call<K extends keyof TProcedures & string>(command: K, ...params: Parameters<TProcedures[K]>):
    Promise<ReturnType<TProcedures[K]>>;
}

export interface RpcServer<TProcedures extends Procedures, TEvents extends Events> {
  register<K extends keyof TProcedures & string>(
    command: K,
    callback: Callable<Promise<ReturnType<TProcedures[K]>>, Parameters<TProcedures[K]>>,
  ): void;
  trigger<K extends keyof TEvents>(event: K, payload: TEvents[K]): void;
}

class Dummy {}

export abstract class Rpc<
    TRemoteProcedures extends Procedures,
    TRemoteEvents extends Events,
    TProcedures extends Procedures,
    TEvents extends Events,
  >
  extends EmitterMixin<typeof Dummy, Events>(Dummy)
  implements RpcClient<TRemoteProcedures, TRemoteEvents>, RpcServer<TProcedures, TEvents>
{
  private calls = new Map<string, [Callable, Callable]>();
  private callbacks = new Map<keyof TProcedures, Callable<Promise<any>, any>>();

  private generateId() {
    let id: string;
    do {
      id = `${Math.random()}`.slice(2);
    } while (this.calls.has(id));

    return id;
  }

  call<K extends keyof TRemoteProcedures & string>(command: K, ...payload: Parameters<TRemoteProcedures[K]>) {
    return new Promise<ReturnType<TRemoteProcedures[K]>>((resolve, reject) => {
      const id = this.generateId();

      this.calls.set(id, [resolve, reject]);
      this.send({ id, command, payload, type: TYPE_REQUEST });
    });
  }

  register<K extends keyof TProcedures & string>(
    command: K,
    callback: Callable<Promise<ReturnType<TProcedures[K]>>, Parameters<TProcedures[K]>>,
  ) {
    this.callbacks.set(command, callback);
  }

  trigger<K extends keyof TEvents>(event: K & string, payload: TEvents[K]) {
    this.send({ event, payload, type: TYPE_EVENT });
  }

  protected process(message: Message) {
    switch (message?.type) {
      case TYPE_EVENT:
        this.processEvent(message);
        break;
      case TYPE_RESPONSE:
        this.processResponse(message);
        break;
      case TYPE_REQUEST:
        this.processRequest(message);
        break;
    }
  }

  private processEvent(event: Event) {
    this.emit(event.event, event.payload);
  }

  private processResponse(response: Response) {
    if (!this.calls.has(response.id)) {
      return;
    }

    const [resolve, reject] = this.calls.get(response.id)!;
    this.calls.delete(response.id);

    if (response.state === STATE_REJECTED) {
      return void reject(response.result);
    }

    resolve(response.result);
  }

  private async processRequest(request: Request) {
    const callback = this.callbacks.get(request.command);

    if (!callback) {
      return;
    }

    try {
      return this.send({
        type: TYPE_RESPONSE,
        id: request.id,
        state: STATE_FULFILLED,
        result: await callback(...request.payload),
      });
    } catch (result) {
      return this.send({
        result,
        type: TYPE_RESPONSE,
        id: request.id,
        state: STATE_REJECTED,
      });
    }
  }

  protected abstract send(message: Message): void;
}
