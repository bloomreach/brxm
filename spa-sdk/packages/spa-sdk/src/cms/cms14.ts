/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { injectable, inject } from 'inversify';
import { EventBusService, EventBus } from './events';
import { CmsOptions, Cms } from './cms';

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

declare global {
  interface Window {
    SPA?: SpaApi;
  }
}

interface SpaApi {
  init(api: CmsApi): void;
  renderComponent(id: string, properties: object): void;
}

interface CmsApi {
  sync(): void;
}

@injectable()
export class Cms14Impl implements Cms {
  private api?: CmsApi;
  private postponed: Function[] = [];

  constructor(@inject(EventBusService) protected eventBus: EventBus) {}

  private async flush() {
    this.postponed
      .splice(0)
      .forEach(task => task());
  }

  private postpone<T extends (...args: any[]) => any>(task: T) {
    return (...args: Parameters<T>) => {
      if (this.api) {
        return task.apply(this, args);
      }

      this.postponed.push(task.bind(this, ...args));
    };
  }

  initialize({ window = GLOBAL_WINDOW }: CmsOptions) {
    if (this.api || !window || window.SPA) {
      return;
    }

    this.eventBus.on('page.ready', this.postpone(this.sync));

    window.SPA = {
      init: this.onInit.bind(this),
      renderComponent: this.onRenderComponent.bind(this),
    };
  }

  protected onInit(api: CmsApi) {
    this.api = api;
    this.flush();
  }

  protected onRenderComponent(id: string, properties: object) {
    this.eventBus.emit('cms.update', { id, properties });
  }

  protected sync() {
    this.api!.sync();
  }
}
