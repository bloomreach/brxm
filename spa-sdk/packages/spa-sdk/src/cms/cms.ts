/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { Events, CmsUpdateEvent } from '../events';
import { RpcClient, RpcServer, Procedures } from './rpc';

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

export interface CmsOptions {
  /**
   * The window reference for the CMS integration.
   * By default the global window object will be used.
   */
  window?: Window;
}

/**
 * CMS integration layer.
 */
export interface Cms {
  /**
   * Initializes integration with the CMS.
   * @param options The CMS integration options.
   */
  initialize(options: CmsOptions): void;
}

interface CmsProcedures extends Procedures {
  sync(): void;
}

interface CmsEvents {
  update: CmsUpdateEvent;
}

interface SpaProcedures extends Procedures {
}

interface SpaEvents {
  ready: never;
}

export class CmsImpl implements Cms {
  private window?: Window;

  constructor(
    protected eventBus: Typed<Events>,
    protected rpcClient: RpcClient<CmsProcedures, CmsEvents>,
    protected rpcServer: RpcServer<SpaProcedures, SpaEvents>,
  ) {
    this.onPageReady = this.onPageReady.bind(this);
    this.onUpdate = this.onUpdate.bind(this);
    this.inject = this.inject.bind(this);
  }

  initialize({ window = GLOBAL_WINDOW }: CmsOptions) {
    if (this.window === window) {
      return;
    }

    this.window = window;
    this.notifyOnReady();
    this.eventBus.on('page.ready', this.onPageReady);
    this.rpcClient.on('update', this.onUpdate);
    this.rpcServer.register('inject', this.inject);
  }

  private notifyOnReady() {
    const notify = () => this.rpcServer.trigger('ready', undefined as never);
    const onStateChange = () => {
      if (this.window!.document!.readyState === 'loading') {
        return;
      }

      notify();
      this.window!.document!.removeEventListener('readystatechange', onStateChange);
    };

    if (this.window?.document?.readyState !== 'loading') {
      return notify();
    }

    this.window?.document?.addEventListener('readystatechange', onStateChange);
  }

  protected onPageReady() {
    this.rpcClient.call('sync');
  }

  protected onUpdate(event: CmsUpdateEvent) {
    this.eventBus.emit('cms.update', event);
  }

  protected inject(resource: string) {
    if (!this.window?.document) {
      return Promise.reject(new Error('SPA document is not ready.'));
    }

    return new Promise((resolve, reject) => {
      const script = this.window!.document.createElement('script');

      script.type = 'text/javascript';
      script.src = resource;
      script.addEventListener('load', () => resolve());
      script.addEventListener('error', () => reject(new Error(`Failed to load resource '${resource}'.`)));
      this.window!.document.body.appendChild(script);
    });
  }
}
