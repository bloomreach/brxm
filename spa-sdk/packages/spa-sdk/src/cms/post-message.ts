/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { injectable } from 'inversify';
import { isMatched } from '../url';
import { Events, Message, Procedures, Rpc } from './rpc';

export const PostMessageService = Symbol.for('PostMessageService');

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

export interface PostMessageOptions {
  /**
   * CMS origin to verify a message sender.
   */
  origin: string;

  /**
   * The window reference for the CMS integration.
   * By default the global window object will be used.
   */
  window?: Window;
}

@injectable()
export class PostMessage<
  TRemoteProcedures extends Procedures = Procedures,
  TRemoteEvents extends Events = Events,
  TProcedures extends Procedures = Procedures,
  TEvents extends Events = Events,
> extends Rpc<TRemoteProcedures, TRemoteEvents, TProcedures, TEvents> {
  private origin?: string;

  private window?: Window;

  constructor() {
    super();
    this.onMessage = this.onMessage.bind(this);
  }

  initialize({ origin, window = GLOBAL_WINDOW }: PostMessageOptions) {
    this.window?.removeEventListener('message', this.onMessage, false);
    this.origin = origin;
    this.window = window;
    this.window?.addEventListener('message', this.onMessage, false);
  }

  protected send(message: Message) {
    if (this.origin) {
      this.window?.parent?.postMessage(message, this.origin);
    }
  }

  private onMessage(event: MessageEvent) {
    if (!event.data || !isMatched(event.origin, this.origin)) {
      return;
    }

    this.process(event.data);
  }
}
