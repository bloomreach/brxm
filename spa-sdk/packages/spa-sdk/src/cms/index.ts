/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ContainerModule } from 'inversify';
import { CmsImpl, CmsService } from './cms';
import { Cms14Impl } from './cms14';
import { PostMessageService, PostMessage } from './post-message';
import { RpcClientService, RpcServerService } from './rpc';

export function CmsModule() {
  return new ContainerModule((bind) => {
    bind(PostMessageService).to(PostMessage).inSingletonScope();
    bind(RpcClientService).toService(PostMessageService);
    bind(RpcServerService).toService(PostMessageService);
    bind(CmsService).to(CmsImpl).inSingletonScope().whenTargetIsDefault();
    bind(CmsService).to(Cms14Impl).inSingletonScope().whenTargetNamed('cms14');
  });
}

export { CmsOptions, CmsService, Cms } from './cms';
export { PostMessageOptions, PostMessageService, PostMessage } from './post-message';
export { RpcClientService, RpcClient, RpcServerService, RpcServer } from './rpc';
