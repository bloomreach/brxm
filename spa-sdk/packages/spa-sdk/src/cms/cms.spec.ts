/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { EventBus, Events } from './events';
import { CmsImpl } from './cms';
import { RpcClient, RpcServer } from './rpc';

describe('CmsImpl', () => {
  let cms: CmsImpl;
  let eventBus: EventBus;
  let rpcClient: jest.Mocked<RpcClient<any, any>>;
  let rpcServer: jest.Mocked<RpcServer<any, any>>;

  beforeEach(() => {
    eventBus = new Typed<Events>();
    rpcClient = {
      call: jest.fn(),
      on: jest.fn(),
      off: jest.fn(),
    };
    rpcServer = {
      register: jest.fn(),
      trigger: jest.fn(),
    };
    cms = new CmsImpl(eventBus, rpcClient, rpcServer);
  });

  describe('initialize', () => {
    const getReadyState = jest.fn();

    beforeEach(() => {
      Object.defineProperty(document, 'readyState', {
        get: getReadyState,
      });
    });

    it('should trigger ready event right away', () => {
      getReadyState.mockReturnValueOnce('interactive');
      cms.initialize({ window });

      expect(rpcServer.trigger).toHaveBeenCalledWith('ready', undefined);
    });

    it('should trigger ready event only once per window', () => {
      getReadyState.mockReturnValueOnce('interactive');
      cms.initialize({ window });
      cms.initialize({ window });

      expect(rpcServer.trigger).toHaveBeenCalledTimes(1);
    });

    it('should not trigger ready event on state change if the state is still loading', async () => {
      getReadyState.mockReturnValueOnce('loading');
      getReadyState.mockReturnValueOnce('loading');
      cms.initialize({ window });

      document.dispatchEvent(new ProgressEvent('readystatechange'));
      await new Promise(resolve => setTimeout(resolve, 0));

      expect(rpcServer.trigger).not.toHaveBeenCalled();
    });

    it('should trigger ready event on state change', async () => {
      getReadyState.mockReturnValueOnce('loading');
      getReadyState.mockReturnValueOnce('interactive');
      cms.initialize({ window });

      document.dispatchEvent(new ProgressEvent('readystatechange'));
      await new Promise(resolve => setTimeout(resolve, 0));

      expect(rpcServer.trigger).toHaveBeenCalledWith('ready', undefined);
    });

    it('should register inject procedure', async () => {
      cms.initialize({ window });

      expect(rpcServer.register).toHaveBeenCalledWith('inject', expect.any(Function));
    });
  });

  describe('onPageReady', () => {
    it('should process postponed events on initialization', async () => {
      cms.initialize({ window });
      await eventBus.emit('page.ready', {});

      expect(rpcClient.call).toHaveBeenCalledWith('sync');
    });
  });

  describe('onUpdate', () => {
    it('should process postponed events on initialization', async () => {
      cms.initialize({ window });

      expect(rpcClient.on).toHaveBeenCalledWith('update', expect.any(Function));

      const [, onUpdate] = rpcClient.on.mock.calls.pop()!;
      const event = { id: 'id', properties: { a: 'b' } };
      const emitSpy = spyOn(eventBus, 'emit');
      onUpdate(event);

      expect(emitSpy).toHaveBeenCalledWith('cms.update', event);
    });
  });

  describe('inject', () => {
    let inject: (string: any) => Promise<void>;
    const sourceDocument = document;
    const getDocument = jest.fn<Document | undefined, []>(() => sourceDocument);

    beforeEach(() => {
      ([[, inject]] = rpcServer.register.mock.calls);
      cms.initialize({ window });
      Object.defineProperty(window, 'document', { get: getDocument });
    });

    it('should not proceed if there is no window', async () => {
      getDocument.mockReturnValueOnce(undefined);

      await expect(inject('something')).rejects.toThrow('SPA document is not ready.');
    });

    it('should put a script element', async () => {
      inject('url1');

      await new Promise(process.nextTick);

      const script = window.document.querySelector('script[src="url1"]');

      expect(script).not.toBeNull();
      expect(script).toMatchSnapshot();
    });

    it('should resolve a promise on load event', async () => {
      const promise = inject('url2');

      await new Promise(process.nextTick);
      const element = window.document.querySelector('script[src="url2"]');
      element?.dispatchEvent(new Event('load'));

      await expect(promise).resolves.toBeUndefined();
    });

    it('should reject a promise on error event', async () => {
      const promise = inject('url3');

      await new Promise(process.nextTick);
      const element = window.document.querySelector('script[src="url3"]');
      element?.dispatchEvent(new Event('error'));

      await expect(promise).rejects.toThrow("Failed to load resource 'url3'.");
    });
  });
});
