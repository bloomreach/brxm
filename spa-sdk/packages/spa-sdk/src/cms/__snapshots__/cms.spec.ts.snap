// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CmsImpl inject should put a script element 1`] = `
<script
  src="url1"
  type="text/javascript"
/>
`;
